ml_model/model/results/compare_results.py

```diff
@@ -0,0 +1,172 @@
+        header['target'] = target
+
+        ## number of data points
+        ## we consider both hyperparameters obtained for code and code+warnings
+        number_of_best_hyperparameters = model_data.shape[0]
+
+        for metric in metrics.keys():
+            ## compute the mean and median values for metrics for both code and code+warnings
+            c_metric = metrics[metric][0]
+            cw_metric = metrics[metric][1]
+            metric_c_mean = model_data[c_metric].mean() ## since we don't have outliers we can reliably use the mean
+            metric_cw_mean = model_data[cw_metric].mean()
+            
+            ## median values
+            print(model_data.groupby(['model','target'])[metrics[metric]].median())
+            header[f'{c_metric} (median)'] = model_data.groupby(['model','target'])[c_metric].median().values[0]
+            header[f'{cw_metric} (median)'] = model_data.groupby(['model','target'])[cw_metric].median().values[0]
+            
+            ## mean values
+            print(model_data.groupby(['model','target'])[metrics[metric]].mean())
+            header[f'{c_metric} (mean)'] = model_data.groupby(['model','target'])[c_metric].mean().values[0]
+            header[f'{cw_metric} (mean)'] = model_data.groupby(['model','target'])[cw_metric].mean().values[0]
+
+            ## show the mean and median values as dots for both code and code+warnings in the boxplots
+            plt.figure(figsize=(10, 6))
+            plt.boxplot([model_data[c_metric], model_data[cw_metric]], labels=[c_metric, cw_metric])
+            plt.title(f'Target: {target} | Model: {model_name} | #best hyperparameters (data points): {number_of_best_hyperparameters}')
+            plt.ylabel(metric)
+            plt.grid(True)
+            plt.scatter(1, metric_c_mean, marker='o', color='red', s=10)
+            plt.scatter(2, metric_cw_mean, marker='o', color='red', s=10)
+            legend_elements = [Line2D([0], [0], marker='o', color='w', label='mean', markerfacecolor='r', markersize=10)]
+            plt.legend(handles=legend_elements, loc='upper right')
+            plt.savefig(ROOT_PATH + 'results/final-boxplots/' + target + '_' + model_name + '_' + metric +'_boxplot_'+ str(SMOTE_TAG) + '.png')
+            plt.close() 
+
+        with open(ROOT_PATH + 'results/mean_median_'+ str(SMOTE_TAG) + '.csv', 'a') as csv_file:
+            writer = csv.DictWriter(csv_file, fieldnames=header.keys())
+            writer.writerow(header)
+
+
+## draw histogram - to check the normality for diff_f1, diff_precision, diff_recall
+columns = ['diff_precision','diff_recall','diff_f1', 'diff_auc']
+for column in columns:
+    hist = overall_df.hist(column=column)
+    plt.savefig(ROOT_PATH + 'results/histograms/final_hist_' + column +'_'+ str(SMOTE_TAG) + '.png')
+
+## draw histogram for all the input features in a single plot
+df_raw_features = pd.read_csv(ROOT_PATH + 'data/understandability_with_warnings.csv')
```

Comment:
I like the histograms in one file, but I don't know what to say. Nadeeshan, figure it out.